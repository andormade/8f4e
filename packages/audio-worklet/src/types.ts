export interface MidiModuleAddresses {
	moduleId: string;
	noteAddress: number | undefined;
	channelAddress: number | undefined;
	noteOnOffAddress: number | undefined;
	velocityAddress: number | undefined;
}

export interface MidiCCModuleAddresses {
	moduleId: string;
	valueAddress: number | undefined;
	channelAddress: number | undefined;
	selectedCCAddress: number | undefined;
}

export interface RNBOModule {
	patcherId: string;
	paramAdresses: number[];
}
