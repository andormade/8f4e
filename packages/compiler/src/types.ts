export interface CompiledModule {
	functionBody: number[];
	initialMemory: number[];
	memoryAddresses: { address: number; id: string; default?: number }[];
	moduleId: string;
	offset: number;
}

export type MemoryBuffer = Int32Array;

export interface Connection {
	fromModuleId: string;
	fromConnectorId: string;
	toModuleId: string;
	toConnectorId: string;
}

export type RelativeAddressCalculator = (nthWord: number) => number;

export type ModuleGenerator<TConfig = unknown> = (
	moduleId: string,
	offset: RelativeAddressCalculator,
	initialConfig?: TConfig
) => CompiledModule;

export type EngineConfig = Record<string, number | string | EngineConfig[]>;

export interface Engine {
	name: string;
	config: EngineConfig;
}

export type ModuleState = Record<string, any>;

export type ModuleStateExtractor<T> = (memoryBuffer: MemoryBuffer, moduleAddress: number) => T;
export type ModuleStateInserter<T> = (moduleState: T, memoryBuffer: MemoryBuffer, moduleAddress: number) => void;

export interface Module {
	id: string;
	engine: Engine;
	state: ModuleState;
}

export type MemoryAddressLookup = Record<string, number>;
