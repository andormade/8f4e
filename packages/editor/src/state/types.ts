import { CompileOptions, CompiledModuleLookup, MemoryBuffer, DataStructure } from '@8f4e/compiler';
import { Font, SpriteLookups } from '@8f4e/sprite-generator';
import { SpriteLookup } from '@8f4e/2d-engine';

export interface CodeBlock {
	code: string[];
	isOpen: boolean;
	x: number;
	y: number;
	viewport?: Viewport;
	codeBlocks?: CodeBlock[];
}

export interface Size {
	width: number;
	height: number;
}

export interface Position {
	x: number;
	y: number;
}

export type Viewport = {
	x: number;
	y: number;
};

interface ContextMenuButton {
	action?: string;
	close?: boolean;
	payload?: Record<string, unknown>;
	disabled?: boolean;
	divider?: boolean;
	title?: string;
	isSectionTitle?: boolean;
}

interface MenuItemDivider extends ContextMenuButton {
	divider: true;
	title: never;
}

export type ExtendedInstructionSet = 'debug' | 'button' | 'switch' | 'offset' | 'plot' | 'piano';

export type ContextMenuItem = ContextMenuButton | MenuItemDivider;

export type MenuGenerator = (state: State, payload?: any) => ContextMenuItem[] | Promise<ContextMenuItem[]>;

export interface ContextMenu extends Position {
	highlightedItem: number;
	itemWidth: number;
	items: ContextMenuItem[];
	open: boolean;
	menuStack: string[];
}

export interface BuildError {
	lineNumber: number;
	code: number;
	message: string;
	moduleId: string;
}

export interface Runtime {
	runner: 'audioWorklet' | 'webWorker';
	latency: number;
	renderQuantum: number;
}

export interface Compiler {
	codeBuffer: Uint8Array;
	compilationTime: number;
	cycleTime: number;
	isCompiling: boolean;
	lastCompilationStart: number;
	memoryBuffer: MemoryBuffer;
	memoryBufferFloat: Float32Array;
	memoryRef: WebAssembly.Memory;
	timerAccuracy: number;
	compiledModules: CompiledModuleLookup;
	buildErrors: BuildError[];
	compilerOptions: CompileOptions;
	allocatedMemorySize: number;
}

export interface Midi {
	outputs: MIDIOutput[];
	inputs: MIDIInput[];
}

export interface MemoryIdentifier {
	memory: DataStructure;
	showAddress: boolean;
	showEndAddress: boolean;
	bufferPointer: number;
	showBinary: boolean;
}

export interface BufferPlotter {
	width: number;
	height: number;
	x: number;
	y: number;
	minValue: number;
	maxValue: number;
	buffer: MemoryIdentifier;
	bufferLength: MemoryIdentifier | undefined;
}
export interface Switch {
	width: number;
	height: number;
	x: number;
	y: number;
	id: string;
	offValue: number;
	onValue: number;
}

export interface Debugger {
	width: number;
	height: number;
	showAddress: boolean;
	showEndAddress: boolean;
	x: number;
	y: number;
	id: string;
	memory: DataStructure;
	bufferPointer: number;
	showBinary: boolean;
}

export interface Output {
	codeBlock: CodeBlockGraphicData;
	width: number;
	height: number;
	x: number;
	y: number;
	id: string;
	calibratedMax: number;
	calibratedMin: number;
	memory: DataStructure;
}

export interface Input {
	codeBlock: CodeBlockGraphicData;
	width: number;
	height: number;
	x: number;
	y: number;
	id: string;
	wordAlignedAddress: number;
}

export interface PianoKeyboard {
	x: number;
	y: number;
	width: number;
	height: number;
	keyWidth: number;
	pressedKeys: Set<number>;
	pressedKeysListMemory: DataStructure;
	pressedNumberOfKeysMemory: DataStructure;
	startingNumber: number;
}

export interface CodeBlockGraphicData {
	width: number;
	minGridWidth: number;
	height: number;
	code: string[];
	trimmedCode: string[];
	padLength: number;
	codeToRender: number[][];
	codeColors: Array<Array<SpriteLookup | undefined>>;
	gaps: Map<number, { size: number }>;
	cursor: { col: number; row: number; x: number; y: number };
	id: string;
	positionOffsetterXWordAddress?: number;
	positionOffsetterYWordAddress?: number;
	x: number;
	y: number;
	offsetX: number;
	offsetY: number;
	gridX: number;
	gridY: number;
	isOpen: boolean;
	extras: {
		inputs: Map<string, Input>;
		outputs: Map<string, Output>;
		debuggers: Map<string, Debugger>;
		bufferPlotters: Map<string, BufferPlotter>;
		switches: Map<string, Switch>;
		buttons: Map<string, Switch>;
		pianoKeyboards: Map<number, PianoKeyboard>;
		errorMessages: Map<
			number,
			{
				message: string[];
				x: number;
				y: number;
			}
		>;
	};
	viewport: Viewport;
	parent: CodeBlockGraphicData;
	codeBlocks: Set<CodeBlockGraphicData>;
}

export type GraphicHelper = {
	spriteLookups?: SpriteLookups;
	outputsByWordAddress: Map<number, Output>;
	globalViewport: {
		width: number;
		height: number;
		roundedWidth: number;
		roundedHeight: number;
		vGrid: number;
		hGrid: number;
		borderLineCoordinates: {
			top: { startX: number; startY: number; endX: number; endY: number };
			right: { startX: number; startY: number; endX: number; endY: number };
			bottom: { startX: number; startY: number; endX: number; endY: number };
			left: { startX: number; startY: number; endX: number; endY: number };
		};
		center: { x: number; y: number };
	};
	baseCodeBlock: CodeBlockGraphicData;
	activeViewport: CodeBlockGraphicData;
	contextMenu: ContextMenu;
	draggedCodeBlock?: CodeBlockGraphicData;
	selectedCodeBlock?: CodeBlockGraphicData;
	dialog: {
		show: boolean;
		text: string;
		title: string;
		buttons: Array<{
			title: string;
			action: string;
			payload?: any;
		}>;
	};
};

export interface BinaryAsset {
	/** The id of the module that the binary data should be loaded into */
	moduleId?: string;
	/** The id of the memory that the binary data should be loaded into */
	memoryId?: string;
	/** The binary data in base64 format */
	data: string;
	/** The file name of the binary data */
	fileName: string;
}

export interface Project {
	title: string;
	author: string;
	description: string;
	codeBlocks: CodeBlock[];
	viewport: Viewport;
	sampleRate: number;
	binaryAssets: BinaryAsset[];
	audioInputBuffers: {
		moduleId: string;
		memoryId: string;
		channel: number;
		input: number;
	}[];
	audioOutputBuffers: {
		moduleId: string;
		memoryId: string;
		channel: number;
		output: number;
	}[];
}

export interface ExampleModule {
	title: string;
	author: string;
	code: string;
	tests: unknown[];
	category: string;
}

export interface Options {
	isLocalStorageEnabled: boolean;
	isDebugMode: boolean;
	localStorageId: string;
	exampleProjects: Record<string, Project>;
	exampleModules: Record<string, ExampleModule>;
}

export interface EditorSettings {
	colorScheme: string;
	font: Font;
}

export interface State {
	compiler: Compiler;
	runtime: Runtime;
	midi: Midi;
	graphicHelper: GraphicHelper;
	project: Project;
	options: Options;
	editorSettings: EditorSettings;
	compilationTime: number;
}
