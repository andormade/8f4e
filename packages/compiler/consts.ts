export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
