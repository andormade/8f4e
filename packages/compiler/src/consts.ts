export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
export const LOGIC_HIGH = I16_SIGNED_LARGEST_NUMBER;
export const LOGIC_LOW = 0;
