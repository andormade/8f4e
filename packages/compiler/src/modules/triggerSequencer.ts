import {
	createFunctionBody,
	createLocalDeclaration,
	i32const,
	i32load,
	i32store,
	ifelse,
	Instruction,
	localGet,
	localSet,
	Type,
} from 'bytecode-utils';
import { MemoryTypes, ModuleGenerator, ModuleStateExtractor, ModuleStateInserter } from '../types';

export enum Memory {
	ZERO,
	TRIGGER_INPUT_POINTER,
	TRIGGER_PREVIOUS_VALUE,
	OUTPUT,
	PATTERN_POINTER,
	PATTERN_MEMORY_SIZE,
	PATTERN_START,
}

enum Locals {
	TRIGGER_INPUT,
	PATTERN_POINTER,
	__LENGTH,
}

export interface Config {
	maxPatternSizeToAlloc: number;
}

const triggerSequencer: ModuleGenerator<Config> = function (moduleId, offset, { maxPatternSizeToAlloc }) {
	const pattern = new Array(maxPatternSizeToAlloc).fill(0);

	const functionBody = createFunctionBody(
		[createLocalDeclaration(Type.I32, Locals.__LENGTH)],
		[
			...i32const(offset.byte(Memory.OUTPUT)),
			...i32const(0),
			...i32store(),

			...i32const(offset.byte(Memory.PATTERN_POINTER)),
			...i32load(),
			...localSet(Locals.PATTERN_POINTER),

			...i32const(offset.byte(Memory.TRIGGER_INPUT_POINTER)),
			...i32load(),
			...i32load(),
			...localSet(Locals.TRIGGER_INPUT),

			// Determining wether the signal is high
			...i32const(offset.byte(Memory.TRIGGER_PREVIOUS_VALUE)),
			...i32load(),
			...localGet(Locals.TRIGGER_INPUT),
			Instruction.I32_GT_S,
			...ifelse(Type.VOID, [
				// Increment pattern pointer
				...localGet(Locals.PATTERN_POINTER),
				...i32const(4),
				Instruction.I32_ADD,
				...i32const(offset.byte(Memory.PATTERN_MEMORY_SIZE)),
				...i32load(),
				Instruction.I32_REM_S,
				...localSet(Locals.PATTERN_POINTER),

				// Set output to high if pattern is set
				...i32const(offset.byte(Memory.PATTERN_START)),
				...localGet(Locals.PATTERN_POINTER),
				Instruction.I32_ADD,
				...i32load(),
				...i32const(1),
				Instruction.I32_EQ,
				...ifelse(Type.VOID, [...i32const(offset.byte(Memory.OUTPUT)), ...i32const(30000), ...i32store()]),

				// Save pattern pointer
				...i32const(offset.byte(Memory.PATTERN_POINTER)),
				...localGet(Locals.PATTERN_POINTER),
				...i32store(),
			]),

			...i32const(offset.byte(Memory.TRIGGER_PREVIOUS_VALUE)),
			...localGet(Locals.TRIGGER_INPUT),
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		byteAddress: offset.byte(0),
		wordAddress: offset.word(0),
		memoryMap: [
			{ address: Memory.ZERO, default: 0, type: MemoryTypes.PRIVATE },
			{
				id: 'in:trigger',
				address: Memory.TRIGGER_INPUT_POINTER,
				default: offset.byte(Memory.ZERO),
				type: MemoryTypes.INPUT_POINTER,
			},
			{ address: Memory.TRIGGER_PREVIOUS_VALUE, default: 0, type: MemoryTypes.PRIVATE },
			{ id: 'out', address: Memory.OUTPUT, default: 0, type: MemoryTypes.OUTPUT },
			{ address: Memory.PATTERN_POINTER, type: MemoryTypes.PRIVATE, default: 0 },
			{ address: Memory.PATTERN_MEMORY_SIZE, type: MemoryTypes.ARRAY_SIZE, default: maxPatternSizeToAlloc * 4 },
			{
				address: Memory.PATTERN_START,
				type: MemoryTypes.DYNAMIC_ARRAY,
				sizePointer: Memory.PATTERN_MEMORY_SIZE,
				maxSize: maxPatternSizeToAlloc * 4,
				default: pattern,
			},
		],
	};
};

export default triggerSequencer;

interface TriggerSequencerState {
	pattern: boolean[];
}

export const insertState: ModuleStateInserter<TriggerSequencerState> = function (
	moduleState,
	memoryBuffer,
	moduleAddress
) {
	const patternStartAddress = moduleAddress + Memory.PATTERN_START;
	const patternMemorySizeAddress = moduleAddress + Memory.PATTERN_MEMORY_SIZE;

	if (moduleState.pattern.length === 0) {
		memoryBuffer[patternMemorySizeAddress] = 1;
		memoryBuffer[patternStartAddress] = 0;
		return;
	}

	memoryBuffer[patternMemorySizeAddress] = moduleState.pattern.length;

	moduleState.pattern.forEach((item, index) => {
		memoryBuffer[patternStartAddress + index * memoryBuffer.BYTES_PER_ELEMENT] = item ? 1 : 0;
	});
};

export const extractState: ModuleStateExtractor<TriggerSequencerState> = function (memoryBuffer, moduleAddress) {
	const patternStartAddress = moduleAddress + Memory.PATTERN_START;
	const patternMemorySizeAddress = moduleAddress + Memory.PATTERN_MEMORY_SIZE;

	const patternMemorySize = memoryBuffer[patternMemorySizeAddress] / memoryBuffer.BYTES_PER_ELEMENT;
	const pattern = Array.from(memoryBuffer.slice(patternStartAddress, patternStartAddress + patternMemorySize));

	return { pattern: pattern.map(item => item === 1) };
};
