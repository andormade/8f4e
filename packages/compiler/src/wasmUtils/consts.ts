export const WASM_MEMORY_PAGE_SIZE = 65536;
