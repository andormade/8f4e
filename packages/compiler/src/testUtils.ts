import wabt from 'wabt';

import {
	createCodeSection,
	createExportSection,
	createFunctionExport,
	createFunctionSection,
	createFunctionType,
	createImportSection,
	createMemoryImport,
	createTypeSection,
} from './wasmUtils/sectionHelpers';
import { FunctionBody } from './wasmUtils/typeHelpers';
import { CompiledModule, TestModule } from './types';
import { WORD_LENGTH } from './consts';

import { compileModules, getInitialMemory } from '.';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

export function createSingleFunctionWASMProgram(functionBody: FunctionBody): Uint8Array {
	return Uint8Array.from([
		...HEADER,
		...VERSION,
		...createTypeSection([createFunctionType([], [])]),
		...createImportSection([createMemoryImport('js', 'memory')]),
		...createFunctionSection([0x00]),
		...createExportSection([createFunctionExport('test', 0x00)]),
		...createCodeSection([functionBody]),
	]);
}

export function setInitialMemory(memory: DataView, module: CompiledModule): void {
	const initialMemory = getInitialMemory(module);
	for (let i = 0; i < initialMemory.length; i++) {
		if (Number.isInteger(initialMemory[i])) {
			memory.setInt32(i * 4, initialMemory[i], true);
		} else {
			memory.setFloat32(i * 4, initialMemory[i], true);
		}
	}
}

export async function createTestModule(sourceCode: string): Promise<TestModule> {
	let allocatedMemoryForTestData = 0;

	const module: CompiledModule = compileModules([{ code: sourceCode.split('\n') }])[0];
	const program = createSingleFunctionWASMProgram(module.functionBody);
	const memoryRef = new WebAssembly.Memory({ initial: 1 });
	const dataView = new DataView(memoryRef.buffer);
	const memoryBuffer = new Int32Array(memoryRef.buffer);
	let instance: WebAssembly.Instance | undefined;
	let wat = '';

	try {
		const webAssemblyInstantiatedSource = await WebAssembly.instantiate(program, {
			js: {
				memory: memoryRef,
			},
		});
		instance = webAssemblyInstantiatedSource.instance;

		wat = await new Promise(resolve => {
			wabt().then(_wabt => {
				const module = _wabt.readWasm(program, {});
				resolve(module.toText({}));
			});
		});
	} catch (error) {
		console.log(error);
	}

	const reset = () => {
		setInitialMemory(dataView, module);

		// Clear the test data that was out of the module's scope.
		for (let i = module.memoryWordSize; i < module.memoryWordSize + allocatedMemoryForTestData; i++) {
			memoryBuffer[i] = 0;
		}
		allocatedMemoryForTestData = 0;
	};

	reset();

	const test = (instance?.exports.test ||
		function () {
			return;
		}) as CallableFunction;

	const memoryGet = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);

			if (!memoryItem) {
				return;
			}

			if (memoryItem.isInteger) {
				return dataView.getInt32(memoryItem.byteAddress, true);
			} else {
				return dataView.getFloat32(memoryItem.byteAddress, true);
			}
		}

		return memoryBuffer[address];
	};

	const memorySet = (address: string | number, value: number | number[]): void => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			if (typeof value === 'number') {
				if (Number.isInteger(value)) {
					dataView.setInt32(memoryItem.byteAddress, value, true);
				} else {
					dataView.setFloat32(memoryItem.byteAddress, value, true);
				}
			} else {
				for (let i = 0; i < value.length; i++) {
					if (Number.isInteger(value[i])) {
						dataView.setInt32(memoryItem.byteAddress + i * 4, value[i], true);
					} else {
						dataView.setFloat32(memoryItem.byteAddress + i * 4, value[i], true);
					}
				}
			}
			return;
		}

		if (typeof value === 'number') {
			if (Number.isInteger(value)) {
				dataView.setInt32(address * 4, value, true);
			} else {
				dataView.setFloat32(address * 4, value, true);
			}
		} else {
			for (let i = 0; i < value.length; i++) {
				if (Number.isInteger(value[i])) {
					dataView.setInt32(address * 4 + i * 4, value[i], true);
				} else {
					dataView.setFloat32(address * 4 + i * 4, value[i], true);
				}
			}
		}
	};

	const getByteAddress = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			return memoryItem.byteAddress;
		}

		return address * WORD_LENGTH;
	};

	const allocMemoryForPointer = (address: string | number): number => {
		const firstFreeAddress = module.memoryWordSize + allocatedMemoryForTestData;
		memorySet(address, firstFreeAddress * WORD_LENGTH);
		allocatedMemoryForTestData++;
		return firstFreeAddress;
	};

	// @ts-ignore
	memoryBuffer.get = memoryGet;
	// @ts-ignore
	memoryBuffer.set = memorySet;
	// @ts-ignore
	memoryBuffer.allocMemoryForPointer = allocMemoryForPointer;
	// @ts-ignore
	memoryBuffer.byteAddress = getByteAddress;

	return {
		// @ts-ignore
		memory: memoryBuffer,
		test,
		reset,
		wat,
		program,
		memoryMap: module.memoryMap,
		ast: module.ast,
	};
}
