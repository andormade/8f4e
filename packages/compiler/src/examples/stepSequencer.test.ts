import triggerSequencer from './stepSequencer.asm';

import { WORD_LENGTH } from '../consts';
import { createTestModule } from '../testUtils';
import { TestModule } from '../types';

let testModule: TestModule;

describe('stepSequencer', () => {
	beforeAll(async () => {
		testModule = await createTestModule(triggerSequencer());
	});

	beforeEach(() => {
		testModule.reset();
	});

	test('if the generated ast matches with the snapshot', () => {
		expect(testModule.ast).toMatchSnapshot();
	});

	test('if the wat code matches with the snapshot', () => {
		expect(testModule.wat).toMatchSnapshot();
	});

	test('if the generated memory map matches with the snapshot', () => {
		expect(testModule.memoryMap).toMatchSnapshot();
	});

	test('if the pointer points to the first element of the step array', () => {
		const { test, memory } = testModule;
		memory.set('steps', 69420);
		test();

		expect(memory.get('out')).toBe(69420);
	});

	test("if the pointer doesn't move when there is no trigger pulse", () => {
		const { memory, test } = testModule;

		for (let i = 0; i < 10; i++) {
			expect(memory.get('stepPointer')).toBe(memory.byteAddress('steps'));
			test();
		}
	});

	test('if the pointer moves when a trigger pulse is provided', () => {
		const { test, memory } = testModule;

		const trigger = memory.allocMemoryForPointer('trigger');
		memory.set('stepLength', 15);

		for (let i = 0; i < 15 * WORD_LENGTH; i += WORD_LENGTH) {
			expect(memory.get('stepPointer')).toBe(memory.byteAddress('steps') + i);
			memory.set(trigger, 1);
			test();
			memory.set(trigger, 0);
			test();
		}
	});

	test('if the pointer resets before it would overflow', () => {
		const { test, memory } = testModule;

		const trigger = memory.allocMemoryForPointer('trigger');
		memory.set('stepLength', 3);

		for (let i = 0; i < 10; i += 1) {
			for (let j = 0; j < 3; j++) {
				expect(memory.get('stepPointer')).toBe(memory.byteAddress('steps') + WORD_LENGTH * j);
				memory.set(trigger, 1);
				test();
				memory.set(trigger, 0);
				test();
			}
		}
	});

	test('if the output changes when a trigger pulse is provided', () => {
		const { test, memory } = testModule;

		const trigger = memory.allocMemoryForPointer('trigger');
		memory.set('stepLength', 4);
		memory.set('steps', [69, 70, 71, 72]);

		for (let i = 69; i < 73; i++) {
			test();
			expect(memory.get('out')).toBe(i);
			memory.set(trigger, 1);
			test();
			memory.set(trigger, 0);
		}
	});
});
