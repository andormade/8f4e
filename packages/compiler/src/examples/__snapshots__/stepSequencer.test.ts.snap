// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stepSequencer if the generated ast matches with the snapshot 1`] = `
Array [
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "stepSequencer",
      },
    ],
    "instruction": "module",
    "lineNumber": 1,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "defaultValue",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "memory",
    "lineNumber": 3,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int*",
      },
      Object {
        "type": "identifier",
        "value": "trigger",
      },
      Object {
        "type": "identifier",
        "value": "&defaultValue",
      },
    ],
    "instruction": "memory",
    "lineNumber": 4,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "triggerPreviousValue",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "memory",
    "lineNumber": 5,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "steps",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 16,
      },
    ],
    "instruction": "buffer",
    "lineNumber": 6,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "stepPointer",
      },
      Object {
        "type": "identifier",
        "value": "&steps",
      },
    ],
    "instruction": "memory",
    "lineNumber": 7,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "stepLength",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 4,
      },
    ],
    "instruction": "memory",
    "lineNumber": 8,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "int",
      },
      Object {
        "type": "identifier",
        "value": "out",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "memory",
    "lineNumber": 9,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_stepPointer",
      },
    ],
    "instruction": "local",
    "lineNumber": 11,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "stepPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 13,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_stepPointer",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 14,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "*trigger",
      },
    ],
    "instruction": "push",
    "lineNumber": 16,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "triggerPreviousValue",
      },
    ],
    "instruction": "push",
    "lineNumber": 17,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterThan",
    "lineNumber": 18,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "if",
    "lineNumber": 19,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_stepPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 20,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "WORD_SIZE",
      },
    ],
    "instruction": "push",
    "lineNumber": 21,
  },
  Object {
    "arguments": Array [],
    "instruction": "add",
    "lineNumber": 22,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_stepPointer",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 23,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_stepPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 25,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "stepLength",
      },
    ],
    "instruction": "push",
    "lineNumber": 26,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 4,
      },
    ],
    "instruction": "push",
    "lineNumber": 27,
  },
  Object {
    "arguments": Array [],
    "instruction": "mul",
    "lineNumber": 28,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&steps",
      },
    ],
    "instruction": "push",
    "lineNumber": 29,
  },
  Object {
    "arguments": Array [],
    "instruction": "add",
    "lineNumber": 30,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterOrEqual",
    "lineNumber": 31,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "if",
    "lineNumber": 32,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&steps",
      },
    ],
    "instruction": "push",
    "lineNumber": 33,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_stepPointer",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 34,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 35,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 36,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&stepPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 37,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_stepPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 38,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 39,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&out",
      },
    ],
    "instruction": "push",
    "lineNumber": 41,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_stepPointer",
      },
    ],
    "instruction": "push",
    "lineNumber": 42,
  },
  Object {
    "arguments": Array [],
    "instruction": "load",
    "lineNumber": 43,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 44,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&triggerPreviousValue",
      },
    ],
    "instruction": "push",
    "lineNumber": 46,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "*trigger",
      },
    ],
    "instruction": "push",
    "lineNumber": 47,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 48,
  },
]
`;

exports[`stepSequencer if the generated memory map matches with the snapshot 1`] = `
Map {
  "defaultValue" => Object {
    "byteAddress": 0,
    "default": 0,
    "id": "defaultValue",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "lineNumber": 3,
    "relativeWordAddress": 0,
    "type": "int",
    "wordAddress": 0,
    "wordSize": 1,
  },
  "trigger" => Object {
    "byteAddress": 4,
    "default": 0,
    "id": "trigger",
    "isInteger": true,
    "isPointer": true,
    "isPointingToInteger": true,
    "lineNumber": 4,
    "relativeWordAddress": 1,
    "type": "int*",
    "wordAddress": 1,
    "wordSize": 1,
  },
  "triggerPreviousValue" => Object {
    "byteAddress": 8,
    "default": 0,
    "id": "triggerPreviousValue",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "lineNumber": 5,
    "relativeWordAddress": 2,
    "type": "int",
    "wordAddress": 2,
    "wordSize": 1,
  },
  "steps" => Object {
    "byteAddress": 12,
    "default": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "id": "steps",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "lineNumber": 6,
    "relativeWordAddress": 3,
    "type": "int",
    "wordAddress": 3,
    "wordSize": 16,
  },
  "stepPointer" => Object {
    "byteAddress": 76,
    "default": 12,
    "id": "stepPointer",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "lineNumber": 7,
    "relativeWordAddress": 19,
    "type": "int",
    "wordAddress": 19,
    "wordSize": 1,
  },
  "stepLength" => Object {
    "byteAddress": 80,
    "default": 4,
    "id": "stepLength",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "lineNumber": 8,
    "relativeWordAddress": 20,
    "type": "int",
    "wordAddress": 20,
    "wordSize": 1,
  },
  "out" => Object {
    "byteAddress": 84,
    "default": 0,
    "id": "out",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "lineNumber": 9,
    "relativeWordAddress": 21,
    "type": "int",
    "wordAddress": 21,
    "wordSize": 1,
  },
}
`;

exports[`stepSequencer if the wat code matches with the snapshot 1`] = `
"(module
  (type (;0;) (func))
  (import \\"js\\" \\"memory\\" (memory (;0;) 1))
  (func (;0;) (type 0)
    (local i32)
    i32.const 76
    i32.load
    local.set 0
    i32.const 4
    i32.load
    i32.load
    i32.const 8
    i32.load
    i32.gt_s
    if  ;; label = @1
      local.get 0
      i32.const 4
      i32.add
      local.set 0
      local.get 0
      i32.const 80
      i32.load
      i32.const 4
      i32.mul
      i32.const 12
      i32.add
      i32.ge_s
      if  ;; label = @2
        i32.const 12
        local.set 0
      end
    end
    i32.const 76
    local.get 0
    i32.store
    i32.const 84
    local.get 0
    i32.load
    i32.store
    i32.const 8
    i32.const 4
    i32.load
    i32.load
    i32.store)
  (export \\"test\\" (func 0)))
"
`;
